orttraining/orttraining/core/optimizer/memory_alleviation.h

```diff
@@ -0,0 +1,285 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+#include <charconv>
+#include "core/common/inlined_containers.h"
+#include "core/common/string_utils.h"
+#include "core/optimizer/graph_transformer.h"
+
+namespace onnxruntime {
+
+/**
+@Class MemoryAlleviation
+
+Find recomputable subgraphs and enable according to user configs.
+*/
+
+class MemoryAlleviation : public GraphTransformer {
+ private:
+  using NodeOutputPort = std::pair<const Node*, int>;
+  using ActivationUsedMap = InlinedHashMap<std::string, std::pair<bool, bool>>;
+
+  /**
+   * @brief Level to control allowed operations during subgraph detecting.
+   * Level 0: only allow cheap-to-compute operations.
+   * Level 1: allow more expensive operations.
+   */
+  enum class ProbeLevel {
+    Basic = 0,
+    Advanced = 1,
+  };
+
+  /**
+   * @brief Type of memory reduction techniques.
+   */
+  enum class AlleviationType {
+    None = 0,  // Disabled.
+    Recompute = 1,
+  };
+
+  /**
+   * @brief Type of user config.
+   * type: type of memory reduction techniques.
+   * requested_count: the number of occurrences of a subgraph pattern for alleviation. -1 means apply all.
+   *   One example: if a subgraph pattern is found 3 times, and requested_count is set 2, then the 1st and 2nd subgraph in
+   *   typological order will be applied for alleviation. This is useful to avoid alleviating more memory than needed.
+   */
+  struct UserAlleviationConfig {
```

Comment:
Can we rename this to RecomputeConfig?

Reply:
or SelectorConfig?

Reply:
I just called it UserConfig. :) because it already in context of memory optimization., :)